Orm/Xtensive.Orm/Orm/Model/IndexInfo.cs

```diff
@@ -320,32 +320,41 @@ public override void UpdateState()
       base.UpdateState();
       CreateColumns();
       valueColumns.UpdateState();
-      foreach (IndexInfo baseIndex in underlyingIndexes) {
+      foreach (var baseIndex in underlyingIndexes) {
         baseIndex.UpdateState();
       }
       filter?.UpdateState();
       CreateTupleDescriptors();
 
-      if (!IsPrimary)
+      if (!IsPrimary) {
         return;
+      }
 
-      var system = new List<int>();
+      var keyColumnsCount = keyColumns.Count;
+      var system = new int[keyColumnsCount + 1];
       var lazy = new List<int>();
-      var regular = new List<int>();
+      var regular = new int[Columns.Count - keyColumnsCount];
+
+      var regularIndex = 0;
+      var systemIndex = 0;
 
-      for (int i = 0, count = Columns.Count; i < count; i++) {
+      for (var i = 0; i < Columns.Count; i++) {
```

Comment:
Invoking `.Count` on every iteration would be inefficient

Reply:
done